trajopt_ifopt/src/discrete_collision_constraint.cpp

```diff
@@ -0,0 +1,264 @@
+  // Get current joint values
+  Eigen::VectorXd joint_vals = this->GetVariables()->GetComponent(position_var_->GetName())->GetValues();
+
+  return CalcValues(joint_vals);
+}
+
+// Set the limits on the constraint values
+std::vector<ifopt::Bounds> DiscreteCollisionConstraintIfopt::GetBounds() const { return bounds_; }
+
+void DiscreteCollisionConstraintIfopt::FillJacobianBlock(std::string var_set, Jacobian& jac_block) const
+{
+  // Only modify the jacobian if this constraint uses var_set
+  if (var_set == position_var_->GetName())
+  {
+    // Get current joint values
+    VectorXd joint_vals = this->GetVariables()->GetComponent(position_var_->GetName())->GetValues();
+
+    CalcJacobianBlock(joint_vals, jac_block);
+  }
+}
+
+Eigen::VectorXd DiscreteCollisionConstraintIfopt::CalcValues(const Eigen::Ref<const Eigen::VectorXd>& joint_vals) const
+{
+  Eigen::VectorXd err = Eigen::VectorXd::Zero(1);
+
+  // Check the collisions
+  tesseract_collision::ContactResultVector dist_results;
+  collision_evaluator_->CalcCollisions(joint_vals, dist_results);
+
+  if (dist_results.empty())
+    return err;
+
+  switch (gradient_method_)
+  {
+    case GradientCombineMethod::SUM:
+    {
+      for (tesseract_collision::ContactResult& dist_result : dist_results)
+      {
+        double dist = collision_evaluator_->GetCollisionConfig().collision_margin_data.getPairCollisionMargin(
+            dist_result.link_names[0], dist_result.link_names[1]);
+        double coeff = collision_evaluator_->GetCollisionConfig().collision_coeff_data.getPairCollisionCoeff(
+            dist_result.link_names[0], dist_result.link_names[1]);
+        err[0] += std::max<double>(((dist - dist_result.distance) * coeff), 0.);
+      }
+      break;
+    }
+    case GradientCombineMethod::WEIGHTED_SUM:
+    {
+      for (tesseract_collision::ContactResult& dist_result : dist_results)
+      {
+        double dist = collision_evaluator_->GetCollisionConfig().collision_margin_data.getPairCollisionMargin(
+            dist_result.link_names[0], dist_result.link_names[1]);
+        double coeff = collision_evaluator_->GetCollisionConfig().collision_coeff_data.getPairCollisionCoeff(
+            dist_result.link_names[0], dist_result.link_names[1]);
+        err[0] += std::max<double>((std::pow(dist - dist_result.distance, 2) * coeff), 0.);
+      }
+      break;
+    }
+    case GradientCombineMethod::AVERAGE:
+    {
+      for (tesseract_collision::ContactResult& dist_result : dist_results)
+      {
+        double dist = collision_evaluator_->GetCollisionConfig().collision_margin_data.getPairCollisionMargin(
+            dist_result.link_names[0], dist_result.link_names[1]);
+        double coeff = collision_evaluator_->GetCollisionConfig().collision_coeff_data.getPairCollisionCoeff(
+            dist_result.link_names[0], dist_result.link_names[1]);
+        double d = std::max<double>(((dist - dist_result.distance) * coeff), 0.);
+        if (d > err[0])
+          err[0] = d;
```

Comment:
This seems to be a misnomer. It doesn't average the error. It actually takes the max.